chpldoc outputs docs in same order as source files.

thomasvandoren opened the pull request:
Update docs() pass to output the documentation in the same order as
the original source files.

Move several of the string utilities previously in docs.cpp into stringutil.cpp
so they can be used elsewhere in the compiler.

Move the logic for printing a particular symbol or class to actual object. For
example, VarSymbol now has a printDocs(file, tabs) method that prints out
the text or rst content for itself. VarSymbol, FnSymbol, ModSymbol, and
AggregateType have printDocs() methods. VarSymbol, FnSymbol, and
AggregateType have docsDirective() methods, which return the rst directive
or text prefix for the instance.

Update ModuleSymbol with docsName() and addPrefixToName() methods.
addPrefixToName() records the parent module names and docsName()
returns the prefix concatenated with the instance's name. Add method to
AggregateType, printInheritanceDocs(), that prints the "inherited from
<name>" section.

Update VarSymbol with new makeField() method that sets private bool
member 'isField' to true. Update AggregateType::addDeclaration to call
makeField() on VarSymbols. This allows a VarSymbol to detect if it is a
global or a class/record member.

Update BaseAst with two methods and a static const member for printing
"tabs" and symbol descriptions. BaseAST::tabText is a static const string
that is used for indentation. It continues to be 3 spaces. printTabs(file, tabs)
adds the indentation to file. printDocsDescription(docs, file, tabs) takes a
doc string for a symbol, cleans it up using the ltrim() et al methods that
used to be in docs.cpp and prints it to file. printDocsDescription() must take
the docs string as an argument because BaseAST does not define a docs
member.

Add new visitor, AstPrintDocs, which has members for a file handle and
tab count. The enter/visit methods call the respective printDocs() for
VarSymbol, FnSymbol, AggregateType, and ModSymbol. When this visitor
is passed to ModSymbol::accept(), the docs are output in source order.

Update docs() pass to use the printDocs() instance methods and remove
the code that printDocs() methods were based on. For now, I have left in the
functions that output the docs in logical groupings and alphabetical order.
The docs() pass does not call them, but they still work (test it by commenting
out the visitor and uncommenting the printModule(file, mod, mod->name) call
inside docs()). In the future (maybe near/medium ish) we can add flags that
allow the caller to choose which ordering they want, which is why I preserved
them.

Update the class/record signatures to include any inherited symbols. If C
inherits A, its signature will now be "class C : A". Multiple inheritance is
supported by joining all inherited classes in a comma delimited list. E.g. if C
inherits from both B and A, its signature will be "class C: B, A".

Remove the "inherited from super_class_name" and additional inherited
members and methods from class/records. This simplifies the AggregateType
visitor method as it no longer has to realize all the inherited symbols in order to
print the docs for itself. It also avoids the duplicate documentation for overridden
methods and members.

In order to make sure the printDocs() methods exhibited the same behavior,
I ran start_test on test/chpldoc/ using the commented out code for logical
groupings. All the chpldoc/ tests passed, so I am confident this only changes
the ordering and not the formatting of any one symbol.

Update 16 .good files under test/chpldoc/ that changed due to the ordering.
Also, mark all test/chpldoc/compflags/alphabetical/ tests as NOTEST, since
the alphabetical ordering does not work with this change. Also, update tests
that included inheritance to match new output.

### TODO:

* [x] ensure existing formatting is preserved by running start_test over
  test/chpldoc with the old ordering (will test that printDocs() methods output
  same format as before and using visitor is really just a change in ordering).
* [x] update chpldoc/ tests for new source based ordering (saved the
   existing .good files, though, for later).
* [x] full test pass (default config on linux64 and LLVM).

thomasvandoren commented:
@lydia-duncan -- Will you please review?

lydia-duncan commented on compiler/AST/type.cpp:
```diff
@@ -1166,6 +1172,99 @@ Symbol* AggregateType::getField(int i) {
+      }
+    }
+
+    if (superClassNames.empty()) {
```
I think there should be an internal error/assert if the inheritance list contains something that we don't expect, which would cause this check to result in true, since it would mean something went wrong during parsing on our end.

Additionally, by not looking up the parent, we're losing some checks which would otherwise occur (records inheriting from classes and vice versa, trying to inherit from a parent that doesn't exist, etc.).  I don't think these checks are the responsibility of the docs pass, but at least the case of inheriting from a non-existent would cause problems for us in documentation generation should it occur.  Should we have a way of detecting that it happened, either here or in the Sphinx domain?

thomasvandoren replied:
What do you think about adding an else block to the if stmt that calls toUnresolvedSymExpr()? The else block would call INT_FATAL() (or would USR_FATAL() be more appropriate?).

lydia-duncan replied:
That sounds good, and INT_FATAL would be ideal.

lydia-duncan commented on compiler/AST/AstPrintDocs.cpp:
```diff
@@ -0,0 +1,273 @@
+
+bool AstPrintDocs::enterFnSym(FnSymbol* node) {
+  node->printDocs(file, this->tabs);
+  return false;
```
Why does this return false always?

thomasvandoren replied:
We're not documenting anything defined within the scope of a function. Returning false indicates the traversal should not go into the function.

lydia-duncan replied:
Ah, that makes sense.

lydia-duncan replied:
Does that logic hold for all enter methods?  If so, we may want to change the stub implementations you have to false.  For instance, I don't think we should enter any of the loop constructs, like WhileDoStmt or CForLoop, since any functions or modules defined there would not be globally defined symbols (as they'd live inside main())

thomasvandoren replied:
I'll write a test and figure out which enter methods on the visitor should return false.

thomasvandoren replied:
It turns out that it is difficult to stop traversal of all things using the visitor. I updated it to avoid traversing into several things, like while/for/do-while loops and conditionals, and filed a future for the things that are harder to detect, like begin/cobegin stmts and tmp/iterator variables that are created.

lydia-duncan replied:
That sounds like the right decision to me, thanks for looking into it. (The loop structures were really the ones I was most worried about)

lydia-duncan commented on compiler/AST/symbol.cpp:
```diff
@@ -2275,6 +2356,92 @@ bool FnSymbol::isSecondaryMethod() const {
+  this->printTabs(file, tabs);
+  *file << this->docsDirective();
+
+  // Print inline/export.
```
Would you please note why extern is not used here, so some misguided soul doesn't come in, see it's missing, then add it?

lydia-duncan commented on compiler/AST/type.cpp:
```diff
@@ -1166,6 +1172,99 @@ Symbol* AggregateType::getField(int i) {
 
+
+void AggregateType::printDocs(std::ostream *file, unsigned int tabs) {
+  if (this->symbol->hasFlag(FLAG_NO_DOC) || this->isUnion()) {
```
Should probably leave a TODO on documenting unions in the future.

lydia-duncan commented on compiler/AST/symbol.cpp:
```diff
@@ -2275,6 +2356,92 @@ bool FnSymbol::isSecondaryMethod() const {
+  *file << info->text();
+  delete info;
+
+  // Print ref return intent, if one exists.
```
Let's drop the "ref" part of the comment, since we are printing any return intent.

lydia-duncan commented on compiler/AST/symbol.cpp:
```diff
@@ -303,6 +308,82 @@ bool VarSymbol::isParameter() const {
+  }
+
+  // FIXME: Do we want this for non-class globals/configs?
+  //        (thomasvandoren, 2015-02-21)
```
I believe we do. @bradcray, we want default values for globals and configs, right?  With the exception of type aliases.

bradcray replied:
Default values for globals and configs yes.

I'm not sure what default values for type aliases means.  Does it mean if I have:

type age = int(64);

that I'll only see:

type age;

in the documentation?  If that's the case, I think we do need to see the "default value" (which I think is actually the definition of the type).  

Last time we discussed this, I think I believed we were talking about the equivalent of a type's "_defaultOf()" value which is why I was surprised that we supported such a capability.  (I.e., I think I was misremembering the type alias syntax as being more like:

type age: int = 21;

(i.e., I was completely confused).

lydia-duncan replied:
I think the definition of a type is an implementation detail.  The only information you should know about it is what it is intended to be used for, and you'd get that from the comment.  Otherwise, people will depend on it being "basically a whatever", breaking at least some of the purpose of wrapping it in a different name.

bradcray replied:
I think the downside to that approach is that it makes the type difficult to use in any contexts other than methods/functions that take that type.  E.g., a user might want to know:
* Can I write an age()?
* Can I pass/use it somewhere that an int is required?
* Can I cast it to int?  
At present, we rely on the int-ness of age for these capabilities.  If it's a complete black box, it limits the usability of the type.

Moreover, in class/record contexts, knowing that a type has a "default value" (let's call it "default definition" to avoid my previous confusion) is crucial for interpreting the type and knowing how to call it.  That is:

record myRange {
  type idxType = int;
  var low, high, stride: idxType;
}

it's key for me to know that:

var r: myRange();

is a legal definition, i.e., that I don't have to type:

var r: myRange(int);

every time I want to create a range (just as default values for var/const/param fields are similarly useful to know when they have default values).

Given both of these concerns, combined with lack of a case in hand where I'd want to squash the definition of the type, I'd prefer to print type definitions in all cases.

lydia-duncan replied:
I think a type field and a global type have sufficiently different uses to justify including the default value for fields but not for global type declarations.

Are there any types we cannot write?  I thought that was defined by default on all types.  In regard to the other two points - "can I pass it somewhere that an int is required" definitely feels like relying on implementation details for convenience (which should not be encouraged on principle) and "can I cast it to _" can be easily denoted by providing a cast function or a toInt() method.  In either case, failing to say the default value of a type does not prevent the user from relying on implementation details that they discover, it merely serves to make it clear that they shouldn't rely on these implementation details.  For instance, what if we wanted age to now be a class that stored information in addition to a numeral, such as an enum to choose between "days", "months" or "years" (and methods to output the correct number depending on the circumstances)?  In that case, we would not be able to provide such functionality without breaking someone's code and rightly so - we told them it was an int, now it isn't.  Whereas, if the underlying type of age was unknown when it was first introduced, we could switch and say "look, here are these new things you can do with age!" with very few people being the wiser (and only those people that rely on undocumented "features").  Sure, we can get away with changing the type under people now from the argument of "probably no one is relying on it", but as we get to be a more mature language we won't be able to do such things and would have to break the habit anyways.

thomasvandoren replied:
FWIW -- My intention was to add it for globals, but I overlooked the fixme before submitting the PR.

@lydia-duncan Do you have a concrete example?

I'm inclined to enable the default for both globals and fields, including types. As with anything, when we find a case that this does not meet our needs, we can revisit.

lydia-duncan replied:
A concrete example of what?

thomasvandoren replied:
Sorry, I misunderstood your last response. Never mind that.

bradcray replied:
The role of the type alias feature is to provide "a type alias" not a brand new type.  For example, if you wrote:

type age = int;

proc foo(x: age); {
  writeln("In age version");
}

proc foo(x: int) {
  writeln("In int version");
}

var myage: age;
var i: int;

foo(myage);
foo(i);

You do not get a cool program in which ints and ages are handled differently, you get an ambiguity because you've written two functions that both take ints.  What that says to me is that I have to know what this type is in order to be able to use it effectively.  If you want a black box type that the user cannot reason about at all, you should write a record and make everything about it private (i.e., you want an ADT not a type alias) or you should propose adding a new structural subtyping feature to the language.  But as the language is defined today, printing out type aliases without what they are aliasing is going to do more good than harm.

I don't buy the type aliases at module and record/class scope should be treated differently argument any more than module and record/class members should be treated differently.

lydia-duncan replied:
Oooooh, okay.  I guess I was thinking in a much more C typedef manner, but what you're describing is obviously the right way to be thinking about it.  Sorry for taking so long to get there.

bradcray replied:
C/C++ are identical to Chapel in this respect (typedef int age; creates an alias to int, not a new type).  The same overload example above C++-ified has the same issue (and of course wouldn't work in C because there's no overloading).

lydia-duncan replied:
Then you confused me with your argument before.  For example, it is considered bad style to rely on size_t being an unsigned int but defining a function to take one or the other would obviously run into the ambiguity error you've described.  When you want to use a size_t, you say you're taking a size_t.  When you want an int, you say you're taking an int.

lydia-duncan replied:
Returning to the record/class versus module level argument, to me, the purpose of a type field for a class or record is to give it generic behavior but also to unite certain fields under a single name.  The user of the type chooses what generic instantiation they want, it is apparent that some information must be provided on that account, even if the type designer doesn't want you to access the inner fields themselves.  However, there is no such need to understand the innards of a type alias, you merely need to know what actions you can take on it and where it is required.

bradcray replied:
In both languages, it's useful to have type aliases for the benefit of creating symbolic names and have the flexibility of mapping the type to different representations over time or across implementations.  In both languages, it's useful to know what the type actually maps down to so that you can do things with it that are legal and avoid things that are illegal (for example, "What control string should I use when I printf() this size_t?"  "Can I add 1 to this size_t?"  "If I overload a function on this type, will it conflict with other overloads?").

I posit that if C had well-defined int sizes, their interfaces would be defined far more often in terms of base types rather than special types like size_t -- that is, I think most of these type aliases cases are designed to deal with ambiguities in the integer sizes.  I expect Chapel to require a lot fewer typedefs in these types of situations (and more generally, fewer module-scoped typedefs) for the same reason.

But I suggest we continue supporting module-level type aliases, and documenting them, and documenting what they map to in spite of this prediction.

bradcray replied:
Why don't we leave it at this:  I assert that chpldoc should print type alias definitions at both scopes.  The day we find a case where that feels like the wrong choice, we can take this debate back up, ideally somewhere other than the comments section of a pull request.

lydia-duncan replied:
If size_t were to change to an integer of size 128 tomorrow, a program that depended on unsigned ints overflowing at a certain point would break.  A program adds 1 to a size_t would not.  One of these actions depends on the underlying type definition.  The other depends on a property of the type that is necessary for its functionality.  The former case is leaked through telling the default definition, the latter should be presented by the documenter for someone to use the type.  You can get all the latter information through merely having an understanding of what the type *does*, while the former information is only available through knowing what the type *is*.

lydia-duncan replied:
I'm fine with leaving the discussion.  I think I've stated my points as strongly and as clearly as I can, and if that isn't enough to convince, then I'll need to revisit how to communicate it better later, because I am firmly convinced that this is encouraging and enabling bad style.

lydia-duncan commented on compiler/AST/type.cpp:
```diff
@@ -550,6 +550,12 @@ addDeclaration(AggregateType* ct, DefExpr* def, bool tail) {
+
+  if (VarSymbol* var = toVarSymbol(def->sym)) {
+    // Identify VarSymbol as class/record member.
+    var->makeField();
```
I wonder if this isField information would be useful to parts of the compiler outside of the docs pass . . .

thomasvandoren replied:
Is this intended to be a question? If so, the answer is no. Or more specifically, it is not used by anything outside the docs() pass.

lydia-duncan replied:
I'm sure there's some other part of the compiler that would treat a VarSymbol differently based on if it was a field or not and so could reasonably make use of this information when it is present instead of whatever it is doing presently, but it would be difficult to track it down so just using it here is the right decision.

thomasvandoren replied:
Oh ok, I see what you are getting at. That does seem like a hard thing to search for.

lydia-duncan commented on compiler/AST/type.cpp:
```diff
@@ -1166,6 +1172,99 @@ Symbol* AggregateType::getField(int i) {
+/*
+ * Add "inherited from <name>" to the documentation.
+ */
+void AggregateType::printInheritanceDocs(std::ostream *file, unsigned int tabs) {
```
This method is not used by anything.  Do we want to comment it out until we've looked into the design implications?

lydia-duncan commented on compiler/include/symbol.h:
```diff
@@ -383,6 +393,12 @@ class FnSymbol : public Symbol {
   bool            isMethod()                                   const;
   bool            isPrimaryMethod()                            const;
   bool            isSecondaryMethod()                          const;
+  bool            isIterator()                                 const;
```
Similarly, is this a method that could be used elsewhere in the compiler?

thomasvandoren replied:
It _could_ be used elsewhere in the compiler. It's not currently, though. It is similar to isMethod(), which is why I added it and use it in FnSymbol::printDocs() and FnSymbol::docsDirective().

thomasvandoren replied:
Were you suggesting that I find other places that were calling `->hasFlag(FLAG_ITERATOR_FN)` and replacing them with this?

lydia-duncan replied:
Yes, that is what I was suggesting.  Similar to what happened when isMethod() and company were added.

thomasvandoren replied:
This is done.

lydia-duncan commented on modules/standard/IO.chpl:
```diff
@@ -156,18 +156,18 @@ const IOHINT_NONE = 0:c_int;
 const IOHINT_RANDOM = QIO_HINT_RANDOM;
 
-/** SEQUENTAL means expect sequential access. On
+/*  SEQUENTAL means expect sequential access. On
```
SEQUENTAL -> SEQUENTIAL

lydia-duncan commented on test/chpldoc/compflags/alphabetical/NOTEST:
```diff
@@ -0,0 +1,4 @@
+These tests verify the alphabetical ordering option in chpldoc. As of
+2015-02-22, that no longer is supported (though not lost forever).
+Re-able these tests when support is returned.
```
Re-enable?

lydia-duncan commented on test/chpldoc/globals/globalOrdering.good:
```diff
@@ -1,11 +1,11 @@
    var fluctuates: int
       I should be fourth 
+   config const resorted
+      I should be second 
```
With source-code based ordering, these comments are confusing and incorrect.  Would you mind updating them to be correct in either case?  i.e "under functional ordering, I should be second, but under source ordering I remained fourth", or something to that effect.

lydia-duncan commented on test/chpldoc/module/ordering.good:
```diff
@@ -1,6 +1,6 @@
       This proc should be printed after all config vars 
 
+   config var first: string
+      This config var should be printed first 
```
Likewise, here the comments are no longer accurate.  Sorry for writing comments/tests like this :(

lydia-duncan commented on test/release/examples/primers/chpldoc.good:
```diff
@@ -27,21 +27,10 @@ Module: chpldoc
+      Class: Bar : Foo
          
          If a class inherits from another class, those fields, methods, and their
          comments are output in both the super and the subclass
```
Likewise, this comment is also out of date atm.

lydia-duncan commented on compiler/passes/docs.cpp:
```diff
@@ -58,73 +51,34 @@ static int compareClasses(const void *v1, const void* v2) {
+          // get the old category based output (or alphabetical). Note that
+          // this will be restored (hopefully soon)... (thomasvandoren, 2015-02-22)
+          //
+          // printModule(file, mod, 0);
```
I'm sorta feeling like we should just go ahead and add the flag now.  It'd a three line change per flag and would stop the NOTEST creep into the chpldoc test directory.

thomasvandoren replied:
While enabling it is only three lines, supporting multiple output formats adds a lot more complexity. I'm going to leave it out for now, since the goal is to document standard modules for the release with the source based ordering.

thomasvandoren commented:
I believe I have addressed all the feedback by fixing the patch or responding inline. I ran tests again with default linux64, llvm, and verify. Please let me know if you have any further feedback.

lydia-duncan commented:
That's great, thanks Thomas!
